File: Project4/Complex.java
public class Complex {
  public Complex(double norm, double imag) {
    this.norm = norm;
    this.imag = imag;
  }

  public double get_norm() {
    return norm;
  }

  public double get_imag() {
    return imag;
  }

  public Complex Add(Complex x) {
    return new Complex(this.norm + x.norm, this.imag + x.imag);
  }

  public Complex Sub(Complex x) {
    return new Complex(this.norm - x.norm, this.imag - x.imag);
  }

  public Complex Mul(Complex x) {
    double norm = this.norm * x.norm - this.imag * x.imag;
    double imag = this.norm * x.norm + this.imag * x.imag;
    return new Complex(norm, imag);
  }

  public Complex Div(Complex x) {
    Complex res = this.Mul(new Complex(x.norm, -x.imag));
    double abs = x.norm * x.norm + x.imag * x.imag;
    return new Complex(res.norm / abs, res.imag / abs);
  }

  public Complex add(double x) {
    return this.Add(new Complex(x, 0));
  }

  public Complex sub(double x) {
    return this.Sub(new Complex(x, 0));
  }

  public Complex mul(double x) {
    return this.Mul(new Complex(x, 0));
  }

  public Complex div(double x) {
    return this.Div(new Complex(x, 0));
  }

  @Override
  public String toString() {
    Complex x = this;
    if (x.norm == 0 && x.imag == 0) {
      return "0";
    } else if (x.norm == 0) {
      return "" + x.imag + "i";
    } else if (x.imag == 0) {
      return "" + x.norm;
    } else if (x.imag > 0) {
      if (x.imag == 1) {
        return "" + x.norm + " +i";
      } else {
        return "" + x.norm + " + " + x.imag + "i";
      }
    } else {
      if (x.imag == -1) {
        return "" + x.norm + "-i";
      } else {
        return "" + x.norm + x.imag + "i";
      }
    }
  }

  private final double norm;
  private final double imag;

  public static void main(String[] args) {
    Complex x = new Complex(1, 2);
    Complex y = new Complex(3, 4);

    System.out.println(x + " + " + y + " = " + x.Add(y));
    System.out.println(x + " - " + y + " = " + x.Sub(y));
    System.out.println(x + " * " + y + " = " + x.Mul(y));
    System.out.println(x + " / " + y + " = " + x.Div(y));
    System.out.println();
    System.out.println(x + " + " + 5 + " = " + x.add(5));
    System.out.println(x + " - " + 5 + " = " + x.sub(5));
    System.out.println(x + " * " + 5 + " = " + x.mul(5));
    System.out.println(x + " / " + 5 + " = " + x.div(5));
  }
}

File: Project4/complex.py
from __future__ import division

class complex:
  def __init__(self, norm = 0, imag = 0):
    self.norm = norm
    self.imag = imag

  def __add__(self, other):
    if isinstance(other, int):
      return complex(self.norm + other, self.imag)
    elif isinstance(other, complex):
      return complex(self.norm + other.norm, self.imag + other.imag)
    else:
      raise TypeError
  
  def __sub__(self, other):
    if isinstance(other, int):
      return complex(self.norm - other, self.imag)
    elif isinstance(other, complex):
      return complex(self.norm - other.norm, self.imag - other.imag)
    else:
      raise TypeError

  def __mul__(self, other):
    if isinstance(other, int):
      return complex(self.norm * other, self.imag * other)
    elif isinstance(other, complex):
      return complex((self.norm * other.norm) - (self.imag * other.imag), (self.norm * other.imag) + (self.imag * other.norm))
    else:
      raise TypeError

  def __truediv__(self, other):
    if isinstance(other, int):
      return complex(round(self.norm / other, 1), round(self.imag / other, 1))
    elif isinstance(other, complex):
      denom = ((other.norm * other.norm) + (other.imag * other.imag))
      first_num = ((self.norm * other.norm) + (other.imag * other.imag))
      second_num = ((self.imag * other.norm) - (self.norm * other.imag))
      return complex(round(first_num / denom, 1), round(second_num / denom, 1))
    else:
      raise TypeError

  def __floordiv__(self, other):
    if isinstance(other, int):
      return complex(round(self.norm / other, 1), round(self.imag / other, 1))
    elif isinstance(other, complex):
      denom = ((other.norm * other.norm) + (other.imag * other.imag))
      first_num = ((self.norm * other.norm) + (self.imag * other.imag))
      second_num = ((self.imag * other.norm) - (self.norm * other.imag))
      return complex(round(first_num / denom, 1), round(second_num / denom, 1))
    else:
      raise TypeError

  def __div__(self, other):
    if isinstance(other, int):
      return complex(round(self.norm / other, 1), round(self.imag / other, 1))
    elif isinstance(other, complex):
      denom = ((other.norm * other.norm) + (other.imag * other.imag))
      first_num = ((self.norm * other.norm) + (self.imag * other.imag))
      second_num = ((self.imag * other.norm) - (self.norm * other.imag))
      return complex(round(first_num / denom, 1), round(second_num / denom, 1))
    else:
      raise TypeError

  def __radd__(self, other):
    return (self + other)

  def __rsub__(self, other):
    return (complex(other) - self)

  def __rmul__(self, other):
    return (self * other)

  def __rtruediv__(self, other):
    return (complex(other) / self)

  def __rfloordiv__(self, other):
    return (complex(other) / self)

  def __rdiv__(self, other):
    return (complex(other) / self)

  def __float__(self):
    return (float(self.norm) / self.imag)

  def __int__(self):
    return (self.norm / self.imag)

  def __str__(self):
    return ('(' + str(self.norm) + ' + ' + str(self.imag) + "i)")


x = complex(1, 2)
y = complex(3, 4)

print('%s + %s = %s' % (x, y, x + y))
print('%s - %s = %s' % (x, y, x - y))
print('%s * %s = %s' % (x, y, x * y))
print('%s / %s = %s' % (x, y, x / y))
print()
print('%s + %i = %s' % (x, 5, x + 5))
print('%s - %i = %s' % (x, 5, x - 5))
print('%s * %i = %s' % (x, 5, x * 5))
print('%s / %i = %s' % (x, 5, x / 5))
print()
print('%i + %s = %s' % (5, x, 5 + x))
print('%i - %s = %s' % (5, x, 5 - x))
print('%i * %s = %s' % (5, x, 5 * x))
print('%i / %s = %s' % (5, x, 5 / x))